diff --git a/drivers/platform/x86/ideapad-laptop.c b/drivers/platform/x86/ideapad-laptop.c
index 784326bd7..33f50dc8f 100644
--- a/drivers/platform/x86/ideapad-laptop.c
+++ b/drivers/platform/x86/ideapad-laptop.c
@@ -869,7 +869,7 @@ static void dytc_profile_refresh(struct ideapad_private *priv)
 
 static int ideapad_dytc_profile_init(struct ideapad_private *priv)
 {
-	int err, dytc_version;
+	int err;
 	unsigned long output;
 
 	if (!priv->features.dytc)
@@ -884,10 +884,6 @@ static int ideapad_dytc_profile_init(struct ideapad_private *priv)
 	if (!test_bit(DYTC_QUERY_ENABLE_BIT, &output))
 		return -ENODEV;
 
-	dytc_version = (output >> DYTC_QUERY_REV_BIT) & 0xF;
-	if (dytc_version < 5)
-		return -ENODEV;
-
 	priv->dytc = kzalloc(sizeof(*priv->dytc), GFP_KERNEL);
 	if (!priv->dytc)
 		return -ENOMEM;
