diff --git a/drivers/platform/x86/ideapad-laptop.c b/drivers/platform/x86/ideapad-laptop.c
index abd0c81d6..bda9a4275 100644
--- a/drivers/platform/x86/ideapad-laptop.c
+++ b/drivers/platform/x86/ideapad-laptop.c
@@ -873,6 +873,13 @@ static void dytc_profile_refresh(struct ideapad_private *priv)
 }
 
 static const struct dmi_system_id ideapad_dytc_v4_allow_table[] = {
+	{
+		/* IdeaPad 3 15ADA05 */
+		.matches = {
+			DMI_MATCH(DMI_SYS_VENDOR, "LENOVO"),
+			DMI_MATCH(DMI_PRODUCT_NAME, "81W1")
+		}
+	},
 	{
 		/* Ideapad 5 Pro 16ACH6 */
 		.matches = {
